### PR TITLE
Introduce ostree kernel initramfs

### PR DESCRIPTION
**Pull request by liuming50**

### CODE REVIEW

**Comment by liuming50** [comment]
Changes in patch set 2:

1 Do not install ostree-initramfs if fitimage is being used.

**Comment by liuming50** [comment]
Changes in patch set 3:

1 Rebase to the latest tip of master branch.

File: classes/image_types_ostree.bbclass [hunk 1]
```diff
@@ -131,27 +130,6 @@ IMAGE_CMD_ostree () {
 
     ln -sf ../var/usrlocal usr/local
 
-    if [ "${KERNEL_IMAGETYPE}" = "fitImage" ]; then
-        # this is a hack for ostree not to override init= in kernel cmdline -
```
Comment (ricardosalveti) [inline classes/image_types_ostree.bbclass]: Did you check if this hack/workaround is also not needed when migrating to ostree-kernel? The empty initramfs when using fit is needed in order for ostree to assume that the system got a initramfs in place (part of fit, in this case).

Comment (liuming50) [reply classes/image_types_ostree.bbclass]: Not yet, only verified it on a arm machine with devicetrees and initramfs, will do that and update the comment, see "WIP" for now.

**Comment by patrickvacek** [comment]
Hello @liuming50, thanks for the submission. I'm still trying to understand the goal here a bit better. I see some explanation in the commit message, but can explain the changes and benefits in more detail?

**Comment by liuming50** [comment]
Hi, @patrickvacek, this patch mainly aims to adapt to the new ostree implementation about how it stores kernel, devicetrees and initramfs, which is in /lib/modules/${KERNEL_VERSION}, the major benefit could be a easier deployment.

**Comment by patrickvacek** [comment]
Is this still WIP and awaiting further updates, or is it ready for review and testing?

**Comment by liuming50** [comment]
Hi, Patrick:

Sorry, have not managed to test that pull request with fitimage.

//Ming Liu

Patrick Vacek <notifications@github.com> 於 2020年2月26日 週三 下午12:08寫道：

File: classes/image_types_ostree.bbclass (hunk 1 above)
Comment (liuming50) [reply classes/image_types_ostree.bbclass:138]: Verified it works for both fitimage and uImage.

**Comment by liuming50** [comment]
Changes in patch set V4:

1 Fixed the conflict with the latest master.

**Comment by patrickvacek** [comment]
I tried running the latest version on our CI, but unfortunately the RPi build fails with this:
```
 ERROR: ostree-kernel-initramfs-0.0.1-r0 do_install: Execution of '/builds/olp/edge/ota/connect/client/meta-updater/build-core-image-minimal-rpi/tmp/work/raspberrypi3-poky-linux-gnueabi/ostree-kernel-initramfs/0.0.1-r0/temp/run.do_install.16344' failed with exit code 1:
cp: cannot stat '/builds/olp/edge/ota/connect/client/meta-updater/build-core-image-minimal-rpi/tmp/deploy/images/raspberrypi3/fitImage-initramfs-ostree-image-raspberrypi3-': No such file or directory
```
Can you tell what is going wrong?

**Comment by liuming50** [comment]
> I tried running the latest version on our CI, but unfortunately the RPi build fails with this:
> 
> ```
>  ERROR: ostree-kernel-initramfs-0.0.1-r0 do_install: Execution of '/builds/olp/edge/ota/connect/client/meta-updater/build-core-image-minimal-rpi/tmp/work/raspberrypi3-poky-linux-gnueabi/ostree-kernel-initramfs/0.0.1-r0/temp/run.do_install.16344' failed with exit code 1:
> cp: cannot stat '/builds/olp/edge/ota/connect/client/meta-updater/build-core-image-minimal-rpi/tmp/deploy/images/raspberrypi3/fitImage-initramfs-ostree-image-raspberrypi3-': No such file or directory
> ```
> 
> Can you tell what is going wrong?

Will check this, I verified the patches in another build environment but not the official meta-updater one, will do that.

File: recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_0.0.1.bb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    install -d $kerneldir
+
+    cp ${DEPLOY_DIR_IMAGE}/${OSTREE_KERNEL} $kerneldir/vmlinuz
+
+    if [ "${KERNEL_IMAGETYPE}" != "fitImage" ]; then
+        if [ -n "${INITRAMFS_IMAGE}" ]; then
+            cp ${DEPLOY_DIR_IMAGE}/${INITRAMFS_IMAGE}-${MACHINE}.${INITRAMFS_FSTYPES} $kerneldir/initramfs.img
+        fi
+
+        if [ "${OSTREE_DEPLOY_DEVICETREE}" = "1"  ] && [ -n "${KERNEL_DEVICETREE}" ]; then
```
Comment (patrickvacek) [inline recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_0.0.1.bb:42]: This should get updated to the `if [ ${@ oe.types.boolean('${OSTREE_DEPLOY_DEVICETREE}')} = True ]` style.

File: recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_0.0.1.bb [hunk 3]
```diff
@@ -0,0 +1,52 @@
+SUMMARY = "Ostree linux kernel, devicetrees and initramfs packager"
+DESCRIPTION = "Ostree linux kernel, devicetrees and initramfs packager"
+SECTION = "kernel"
+LICENSE = "GPLv2"
+LIC_FILES_CHKSUM = "file://${COMMON_LICENSE_DIR}/GPL-2.0;md5=801f80980d171dd6425610833a22dbe6"
+
+# Whilst not a module, this ensures we don't get multilib extended (which would make no sense)
+inherit module-base
+
+PACKAGES = "ostree-kernel ostree-initramfs ostree-devicetrees"
```
Comment (patrickvacek) [inline recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_0.0.1.bb:10]: How do these packages actually get populated? I think I'm missing the connection.

File: recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_0.0.1.bb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            mkdir -p $kerneldir/dtb
+            for dts_file in ${KERNEL_DEVICETREE}; do
+                dts_file_basename=$(basename $dts_file)
+                cp ${DEPLOY_DIR_IMAGE}/$dts_file_basename $kerneldir/dtb/$dts_file_basename
+            done
+        fi
+    fi
+}
+do_install[vardepsexclude] = "KERNEL_VERSION"
+do_install[depends] = "virtual/kernel:do_deploy ${@['${INITRAMFS_IMAGE}:do_image_complete', ''][d.getVar('INITRAMFS_IMAGE') == '']}"
```
Comment (patrickvacek) [inline recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_0.0.1.bb:52]: The latter half of this line uses a syntax that I'm having trouble following. Can you elucidate that for me?

File: recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_0.0.1.bb (hunk 2 above)
Comment (liuming50) [reply recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_0.0.1.bb:42]: Yes, you are right, will change that in V5.

File: recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_0.0.1.bb (hunk 3 above)
Comment (liuming50) [reply recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_0.0.1.bb:10]: These packages are being installed by IMAGE_INSTALL

File: recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_0.0.1.bb (hunk 4 above)
Comment (liuming50) [reply recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_0.0.1.bb:52]: This is another syntax of oe.utils.ifelse or '...if..else'.

Comment (liuming50) [reply recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_0.0.1.bb:52]: I am trying to reproduce the error, will keep you updated.

File: recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_0.0.1.bb (hunk 3 above)
Comment (patrickvacek) [reply recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_0.0.1.bb:10]: I guess I meant, how do you define what the contents are of each of these packages? I'm not seeing how the files that are getting copied/installed below actually get assigned to one or the other of these packages.

Comment (liuming50) [reply recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_0.0.1.bb:10]: I am not sure if I understood you question, but if you meant how the artifacts being packaged, the answer would be by FILES?
```
+FILES_ostree-kernel = "${nonarch_base_libdir}/modules/*/vmlinuz"
+FILES_ostree-initramfs = "${nonarch_base_libdir}/modules/*/initramfs.img"
+FILES_ostree-devicetrees = "${nonarch_base_libdir}/modules/*/dtb/*"
```

**Comment by liuming50** [comment]
@patrickvacek Just pushed a V6 patch, please kindly help me review it.

The changes in V6:

1 Change to oe.types.boolean to check the value of OSTREE_DEPLOY_DEVICETREE.

2 inherit kernel-artifact-names in ostree-kernel-initramfs_0.0.1.bb to get the kernel artifacts' names.

File: recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_0.0.1.bb (hunk 3 above)
Comment (patrickvacek) [reply recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_0.0.1.bb:10]: Ah, perfect, thanks. I was overlooking that bit.

**Comment by patrickvacek** [comment]
Hey @liuming50, my apologies for this taking so long. I'd like to finally test this and merge this. Can you rebase this and resolve the conflict? Seems pretty minor to me.

**Comment by liuming50** [comment]
@patrickvacek Sure, will rebase it

**Comment by liuming50** [comment]
@patrickvacek  Changes in patch set V7:

1 Rebased against the latest master

**Comment by patrickvacek** [comment]
> 1 Rebased against the latest master

Thanks! LGTM and CI passes.

**Comment by liuming50** [comment]
Hi, Patrick:

Is there any change this patch goes to zeus? I could make patches and test
them.

the best,
thank you

Patrick Vacek <notifications@github.com> 於 2020年6月22日 週一 上午9:40寫道：

**Comment by patrickvacek** [comment]
> Is there any change this patch goes to zeus? I could make patches and test them.

Yes! Our next release/backport cycle will be a bit delayed but probably in 2-3 weeks. If you want this in zeus before then, feel free to make a PR.

**Comment by liuming50** [comment]
Thanks for the reply, I will work on the patches.

//Ming Liu

Patrick Vacek <notifications@github.com> 於 2020年6月24日 週三 下午1:49寫道：

**Pull request merged**